File: ws/src/robot_remote/src/task_server.cpp
#include <moveit_core/moveit/utils/moveit_error_code.h>
#include <moveit_ros_planning_interface/moveit/move_group_interface/move_group_interface.h>
#include <rclcpp/logger.hpp>
#include <rclcpp/logging.hpp>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include <memory>
#include <robot_msgs/action/robot_task.hpp>
#include <rclcpp_action/rclcpp_action/server_goal_handle.hpp>
#include <thread>
#include <rclcpp_components/register_node_macro.hpp>
#include <moveit/move_group_interface/move_group_interface.h>

using namespace std::placeholders;
using namespace std::chrono_literals;
namespace robot_remote
{
  class TaskServer : public rclcpp::Node
  {
  public:
    explicit TaskServer(const rclcpp::NodeOptions &options = rclcpp::NodeOptions()) : Node("task_server", options)
    {
      action_server_ = rclcpp_action::create_server<robot_msgs::action::RobotTask>(
          this,
          "task_server",
          std::bind(&TaskServer::goalCallback, this, _1, _2),
          std::bind(&TaskServer::cancelCallback, this, _1),
          std::bind(&TaskServer::acceptedCallback, this, _1));

      RCLCPP_INFO(this->get_logger(), "Starting the Action server");
    }

  private:
    rclcpp_action::Server<robot_msgs::action::RobotTask>::SharedPtr action_server_;

    rclcpp_action::GoalResponse goalCallback(const rclcpp_action::GoalUUID &uuid, std::shared_ptr<const robot_msgs::action::RobotTask::Goal> goal)
    {
      RCLCPP_INFO_STREAM(this->get_logger(), "Received goal request with id:" << goal->task_number);
      return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    };

    void acceptedCallback(const std::shared_ptr<rclcpp_action::ServerGoalHandle<robot_msgs::action::RobotTask>> goal_handle)
    {
      std::thread{std::bind(&TaskServer::execute, this, _1), goal_handle}.detach();
    };

    void execute(const std::shared_ptr<rclcpp_action::ServerGoalHandle<robot_msgs::action::RobotTask>> goal_handle)
    {
      RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Executing goal");
      auto arm_move_group = moveit::planning_interface::MoveGroupInterface(shared_from_this(), "arm");
      auto gripper_move_group = moveit::planning_interface::MoveGroupInterface(shared_from_this(), "gripper");

      std::vector<double> arm_joint_goal;
      std::vector<double> gripper_joint_goal;
      
      if(goal_handle->get_goal()->task_number == 0){
        arm_joint_goal = {0.0, 0.0, 0.0};
        gripper_joint_goal = {-0.7, 0.7};
      }
      else if(goal_handle->get_goal()->task_number == 1){
          arm_joint_goal = {-1.14, -0.35, -0.06};
        gripper_joint_goal = {0.0, 0.0};
      }
      else if(goal_handle->get_goal()->task_number == 2){
          arm_joint_goal = {-1.57, -0.13, -1.0};
        gripper_joint_goal = {-0.0, 0.0};
      }
      else{
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Invalid task Number");
        return;
      }

      bool arm_within_bounds = arm_move_group.setJointValueTarget(arm_joint_goal);
      bool gripper_within_bounds = gripper_move_group.setJointValueTarget(gripper_joint_goal);
    
      if(!arm_within_bounds | !gripper_within_bounds){
        RCLCPP_WARN(rclcpp::get_logger("rclcpp"), "Target position were outside limits");
        return;
      }

      moveit::planning_interface::MoveGroupInterface::Plan arm_plan;
      moveit::planning_interface::MoveGroupInterface::Plan gripper_plan;

      arm_move_group.setPlanningTime(10.0);
      gripper_move_group.setPlanningTime(10.0);

      auto arm_plan_result = arm_move_group.plan(arm_plan);

      bool arm_plan_success = arm_move_group.plan(arm_plan) == moveit::core::MoveItErrorCode::SUCCESS;
      bool gripeer_plan_success = gripper_move_group.plan(gripper_plan) == moveit::core::MoveItErrorCode::SUCCESS;

      RCLCPP_INFO_STREAM(rclcpp::get_logger("rclcpp"), "arm_plan " << arm_plan_success << " Arm bounds " << arm_within_bounds << " Arm plan code: " << arm_plan_result.val << " planning time " << arm_move_group.getPlanningTime());
      RCLCPP_INFO_STREAM(rclcpp::get_logger("rclcpp"), "gripper_plan " << gripeer_plan_success << " Gripper bounds " << gripper_within_bounds);


      if(arm_plan_success && gripeer_plan_success){
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Planner succedd, moving the arm and the gripper");
        arm_move_group.move();
        gripper_move_group.move();
      }
      else{
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "One or more planner failed");
        return;
      }
    auto result = std::make_shared<robot_msgs::action::RobotTask::Result>();
    result->success = true;

    goal_handle->succeed(result);
    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "Goal succeeded");
    };

    rclcpp_action::CancelResponse cancelCallback(std::shared_ptr<rclcpp_action::ServerGoalHandle<robot_msgs::action::RobotTask>> goal_handle)
    {
      RCLCPP_INFO(this->get_logger(), "Received request to cancel the goal");
      auto arm_move_group = moveit::planning_interface::MoveGroupInterface(shared_from_this(), "arm");
      auto gripper_move_group = moveit::planning_interface::MoveGroupInterface(shared_from_this(), "gripper");
      //   (void)goal_handle;
      arm_move_group.stop();
      gripper_move_group.stop();
      return rclcpp_action::CancelResponse::ACCEPT;
    };
  };
}

RCLCPP_COMPONENTS_REGISTER_NODE(robot_remote::TaskServer)

File: ws/src/robot_remote/robot_remote/task_server.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer
from rclpy.action.server import ServerGoalHandle
from robot_msgs.action import RobotTask
import numpy as np
import time
from moveit.planning import MoveItPy
from moveit.core.robot_state import RobotState

class TaskServer(Node):
    def __init__(self):
        super().__init__("task_server")
        self.action_server: ActionServer = ActionServer(self, RobotTask, 'task_server', self.goalCallback)
        self.__bot = MoveItPy(node_name="moveit_py")
        self.bot_arm = self.__bot.get_planning_component("arm")
        self.bot_gripper = self.__bot.get_planning_component("geipper")

    def goalCallback(self, goal_handle: ServerGoalHandle)-> RobotTask.Result | None:
        self.get_logger().info(f"Received goal request with task number {goal_handle.request.task_number}")
        feedback_msg = RobotTask.Feedback()
        arm_state = RobotState(self.__bot.get_robot_model())
        gripper_state = RobotState(self.__bot.get_robot_model())

        arm_joint_goal = []
        gripper_joint_goal = [] 
    
        match goal_handle.request.task_number:
            case 0:
                arm_joint_goal = np.array([0.0, 0.0, 0.0])
                gripper_joint_goal = np.array([-0.7, 0.7]) 
            case 1:
                arm_joint_goal = np.array([-1.14, -0.35, -0.06])
                gripper_joint_goal = np.array([0.0, 0.0])
            case 2:
                arm_joint_goal = np.array([-1.57, -0.13, -1.0])
                gripper_joint_goal = np.array([0.0, 0.0])
            case _:
                self.get_logger().error("Invalid task number")
                return
        
        arm_state.set_joint_group_positions("arm", arm_joint_goal)
        gripper_state.set_joint_group_positions("arm", gripper_joint_goal)

        self.bot_arm.set_start_state_to_current_state()
        self.bot_gripper.set_start_state_to_current_state()

        self.bot_arm.set_goal_state(robot_state=arm_state)
        self.bot_gripper.set_goal_state(robot_state=gripper_state)

        arm_plan_result = self.bot_arm.plan()
        gripper_plan_result = self.bot_gripper.plan()

        if arm_plan_result and gripper_plan_result:
            self.__bot.execute(arm_plan_result.trajectory, controllers=[])
            self.__bot.execute(gripper_plan_result.trajectory, controllers=[])
        else:
            self.get_logger().info("One or more planner failed.")
        
        goal_handle.succeed()
        result = RobotTask.Result()
        result.success = True

        return result


def main():
    rclpy.init()
    task_server = TaskServer()
    rclpy.spin(task_server)
    task_server.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()